app/controllers/planning_applications_controller.rb

```diff
@@ -71,14 +71,37 @@ def withdraw
     redirect_to @planning_application
   end
 
+  def cancel_confirmation
+    render :cancel_confirmation
+  end
+
+  def cancel
+    status = authorize_user_can_update_status(
+        planning_application_params[:status])
+    apply_cancellation(status)
+    @planning_application.update!(cancellation_comment: planning_application_params[:cancellation_comment])
+    if @planning_application.withdrawn? || @planning_application.returned?
+      flash[:notice] = "Application has been cancelled"
+    end
+    redirect_to @planning_application
+  end
+
+  def apply_cancellation(status)
+    if status == "withdrawn"
+      @planning_application.withdraw!
+    elsif status == "returned"
+      @planning_application.return!
+    end
+  end
+
   private
 
   def set_planning_application
     @planning_application = authorize(PlanningApplication.find(params[:id]))
   end
 
   def planning_application_params
```

Comment:
This has made me realize that we're doing any bulk attribute updating in this controller at all? If that's the case, we can get get rid of this method and it's use of strong parameters.